### PR TITLE
Supply Values from Remote Config Files

### PR DESCRIPTION
**Pull request by phoban01**
## Description

This PR adds support for configuration via remote files.

The field `valuesFrom` is now available on the `Configuration` object:

```yaml
apiVersion: delivery.ocm.software/v1alpha1
kind: Configuration
metadata:
  name: podinfo-deployment
  namespace: ocm-system
spec:
  interval: 1m
  sourceRef:
    kind: Localization
    name: podinfo-deployment
  configRef:
    kind: ComponentVersion
    name: podinfo
    resourceRef:
      name: config
  valuesFrom:
    fluxSource:
      sourceRef:
          kind: GitRepository
          name: flux-system
          namespacce: flux-system
      path: "config/values.yaml" # path to the value file in the flux source
      subPath: production.deployment # optional subpath within the yaml file itself
```

Note that currently we only support a single source for configuration values, and merging is not possible.

Signed-off-by: Piaras Hoban <phoban01@gmail.com>

### CODE REVIEW

**Comment by Skarlso** [comment]
>      subPath: production.deployment # optional

I'm not sure it's possible to have nested resources?

**Comment by Skarlso** [comment]
The only problem I can see here is that what if someone chooses a different name for the initial flux repo `GitRepositroy`? how would you communicate that down to the pipeline object / product controller / anything that tries to use valuesFrom?

File: controllers/mutation_reconcile_looper.go [hunk 1]
```diff
@@ -758,3 +765,112 @@ func (m *MutationReconcileLooper) getComponentVersion(ctx context.Context, obj *
...
+		return obj.Values, nil
+	}
+
+	if obj.ValuesFrom.FluxSource != nil {
+		source, err := m.getSource(ctx, obj.ValuesFrom.FluxSource.SourceRef)
+		if err != nil {
+			return nil, fmt.Errorf("could not get values from source: %w", err)
+		}
+
+		tmpDir, err := os.MkdirTemp("", "mutation-controller-")
```
Comment (Skarlso) [inline controllers/mutation_reconcile_looper.go:782]: Forgot error handling here. :)

File: controllers/mutation_reconcile_looper.go [hunk 2]
```diff
@@ -758,3 +765,112 @@ func (m *MutationReconcileLooper) getComponentVersion(ctx context.Context, obj *
...
+		case map[interface{}]interface{}:
+			// Recursively convert nested maps
+			result[key] = convertMap(value)
+		default:
+			result[key] = value
+		}
+	}
+
+	return result
+}
```
Comment (Skarlso) [inline controllers/mutation_reconcile_looper.go:876]:
```suggestion
func convertMap(data map[any]any) map[string]any {
	result := make(map[string]any)

	for k, v := range data {
		key, ok := k.(string)
		if !ok {
			// Handle the case where the key is not a string
			continue
		}

		switch value := v.(type) {
		case map[any]any:
			// Recursively convert nested maps
			result[key] = convertMap(value)
		default:
			result[key] = value
		}
	}

	return result
}
```

Just looks neater. :D

**Comment by phoban01** [comment]
> The only problem I can see here is that what if someone chooses a different name for the initial flux repo `GitRepositroy`? how would you communicate that down to the pipeline object / product controller / anything that tries to use valuesFrom?

I don't think that's a proble that the ocm-controller needs to solve, it just needs to provide the basic functionality. But this has reminded me that we need to index the sourceRef field.

**Comment by phoban01** [comment]
e2e tests fixed

File: api/v1alpha1/mutation_types.go [hunk 3]
```diff
@@ -45,6 +48,28 @@ type MutationSpec struct {
 	Suspend bool `json:"suspend,omitempty"`
 }
 
+// ValuesSource provides access to values from an external Source such as a ConfigMap or GitRepository
+// an optional subpath defines the path within the source from which the values should be resolved.
```
Comment (souleb) [inline api/v1alpha1/mutation_types.go:52]:
```suggestion
// An optional subpath defines the path within the source from which the values should be resolved.
```

**Review by Skarlso — APPROVED** [review]
Tested this extensively.

```
➜  mpas-product-controller-examples k get configuration -A
NAMESPACE     NAME                     READY   SOURCE VERSION   CONFIG VERSION   AGE
mpas-system   backend-configuration    True    1.0.1            1.0.1            3m58s
mpas-system   cache-configuration      True    1.0.1            1.0.1            3m58s
mpas-system   frontend-configuration   True    1.0.1            1.0.1            3m58s
➜  mpas-product-controller-examples k get localization -A
NAMESPACE     NAME                    READY   SOURCE VERSION   CONFIG VERSION   AGE
mpas-system   backend-localization    True    1.0.1            1.0.1            4m23s
mpas-system   cache-localization      True    1.0.1            1.0.1            4m23s
mpas-system   frontend-localization   True    1.0.1            1.0.1            4m23s
➜  mpas-product-controller-examples

```

**Pull request merged**